include/boost/iostreams/filter/zlib.hpp

```diff
@@ -347,7 +347,7 @@ bool zlib_compressor_impl<Alloc>::filter
     int result = xdeflate(flush ? zlib::finish : zlib::no_flush);
     after(src_begin, dest_begin, true);
     zlib_error::check BOOST_PREVENT_MACRO_SUBSTITUTION(result);
-    return result != zlib::stream_end; 
+    return result == zlib::okay; 
```

Comment:
The right way is to return (result == zlib::okay) because when we return true, the symmetric_filter continues filtering. However if there is an error (not throwing exception), it should stop filtering and return immediately. By this fix, there is only 3 values not throwing exception (Z_OK, Z_STREAM_END, Z_BUF_ERROR).
If we decide to add another value (e.g. Z_MEM_ERROR) to this noexception list, the filter function shall return false for that error code too, otherwise the symmetric_filter will enter an infinite loop.
So that returning (result == zlib::okay) is the right way.